File: java/lesson 10/FibonacciNumberRecursive.java
// https://leetcode.com/problems/fibonacci-number/

class Solution {
    public int fib(int n) {
        if (n <= 1) return n;

        Integer[] dp = new Integer[n + 1];
        dp[0] = 0;
        dp[1] = 1;

        return fib(n, dp);
    }

    private int fib(int n, Integer[] dp) {
        if (dp[n] != null) return dp[n];
        dp[n] = fib(n - 1, dp) + fib(n - 2, dp);
        return dp[n];
    }
}

File: csharp/lesson 10/FibonacciNumberRecursive.cs
// https://leetcode.com/problems/fibonacci-number/

public class Solution
{
    public int Fib(int n)
    {
        if (n <= 1) return n;
        return Fib(n, new int[n + 1]);
    }
    
    private int Fib(int n, int[] dp)
    {
        if (n <= 1) return n;
        if (dp[n] > 0) return dp[n];        
        dp[n] = Fib(n - 1, dp) + Fib(n - 2, dp);
        return dp[n];
    }
}
